src/clj/puppetlabs/trapperkeeper/services/watcher/filesystem_watch_core.clj

```diff
@@ -53,31 +53,53 @@
        :changed-path (.. watched-path (resolve (.context event)) (toFile))})))
 
 (defn validate-watch-options!
+  "Validate that the options supplied include a valid Boolean value for key :recursive."
   [options]
-  (when-not (= true (:recursive options))
+  (when-not (instance? Boolean (:recursive options))
     (throw
       (IllegalArgumentException.
-        (trs "Support for non-recursive directory watching not yet implemented")))))
+        (trs "Must pass a boolean value for :recursive (directory watching) option")))))
 
 (defrecord WatcherImpl
-  [watch-service callbacks]
+  [watch-service callbacks recursive]
   Watcher
   (add-watch-dir!
+    [this dir]
+    (let [watched-path (.toPath (fs/file dir))]
+      (if @recursive
+        (DirWatchUtils/registerRecursive watch-service [watched-path])
+        (DirWatchUtils/register watch-service watched-path))))
+
+  (add-watch-dir!
     [this dir options]
     (validate-watch-options! options)
-    (DirWatchUtils/registerRecursive watch-service
-                                     [(.toPath (fs/file dir))]))
+    ;; The value of recursive was already set by `create-watcher`, possibly by specifying an option.
+    ;; We validate that the option supplied to `add-watch-dir!` has the same value.
+    (when-not (= @recursive (:recursive options))
+      (throw
+        (IllegalArgumentException.
+          (trs "Watcher already set to :recursive {0}, cannot change to :recursive {1}"
+            @recursive (:recursive options)))))
+    (watch-protocol/add-watch-dir! this dir))
+
   (add-callback!
     [this callback]
     (swap! callbacks conj callback)))
 
 (defn create-watcher
-  []
-  (map->WatcherImpl
-    {:watch-service (.newWatchService (FileSystems/getDefault))
-     :callbacks (atom [])}))
+  ([]
+   (create-watcher {:recursive true}))
+  ([options]
```

Comment:
It makes it clear which options the function is caring about.

It would matter more if there were a bunch of keys being pulled out of the map, in this case either way is fine.